src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala

```diff
@@ -169,9 +163,50 @@ class TaskReplaceActor(
 }
 
 object TaskReplaceActor {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
   val KillComplete = "^TASK_(ERROR|FAILED|FINISHED|LOST|KILLED)$".r
   val FailedToStart = "^TASK_(ERROR|FAILED|LOST|KILLED)$".r
 
   case object RetryKills
+
+  /** Encapsulates the logic how to get a Restart going */
+  private[upgrade] case class RestartStrategy(nrToKillImmediately: Int, maxCapacity: Int)
+
+  private[upgrade] def computeRestartStrategy(app: AppDefinition, runningTasksCount: Int): RestartStrategy = {
+    // in addition to an app definition which passed validation, we require:
+    require(app.instances > 0, s"instances must be > 0 but is ${app.instances}")
+    require(runningTasksCount >= 0, s"running task count must be >=0 but is $runningTasksCount")
```

Comment:
Not sure if this requirement is valid:
- deploy application with n tasks
- the deployment is successful
- all tasks fail, all restart attempts fail as well
- we deploy a new version at exactly a time all tasks are gone